sdk/python/kfp/components/types/custom_artifact_types.py

```diff
@@ -0,0 +1,201 @@
+
+    Args:
+        obj: The class or module for which to get the fully qualified name.
+
+    Returns:
+        The fully qualified name for the class.
+    """
+    module = obj.__module__
+    name = obj.__qualname__
+    if module is not None:
+        name = module + '.' + name
+    return name
+
+
+def get_import_item_from_annotation_string_and_ann_obj(artifact_str: str,
+                                                       qualname: str) -> str:
+    """Gets the fully qualified names of the module or type to import from the
+    annotation string and the annotation object.
+
+    Args:
+        artifact_str: The type annotation string.
+        qualname: The fully qualified type annotation name as a string.
+
+    Returns:
+        The fully qualified names of the module or type to import.
+    """
+    split_qualname = qualname.split('.')
+    if artifact_str in split_qualname:
+        name_to_import = '.'.join(
+            split_qualname[:split_qualname.index(artifact_str) + 1])
+    else:
+        raise TypeError(
+            f"Module or type name aliases are not supported. You appear to be using an alias in your type annotation: '{qualname}'. This may be due to use of an 'as' statement in an import statement or a reassignment of a module or type to a new name. Reference the module and/or type using the name as defined in the source from which the module or type is imported."
+        )
+    return name_to_import
+
+
+def func_to_annotation_object(func: Callable) -> Dict[str, str]:
+    """Gets a dict of parameter name to annotation object for a function."""
+    signature = inspect.signature(func)
+    return {
+        name: parameter.annotation
+        for name, parameter in signature.parameters.items()
+    }
+
+
+def func_to_root_annotation_symbols_for_artifacts(
+        func: Callable) -> Dict[str, str]:
+    """Gets a map of parameter name to the root symbol used to reference an
+    annotation for a function.
+
+    For example, the annotation `type_module.Artifact` has the root
+    symbol `type_module` and the annotation `Artifact` has the root
+    symbol `Artifact`.
+    """
+    module_node = ast.parse(
+        component_factory._get_function_source_definition(func))
+    args = module_node.body[0].args.args
+
+    def convert_ast_arg_node_to_param_annotation_dict_for_artifacts(
+            args: List[ast.arg]) -> Dict[str, str]:
+        """Converts a list of ast.arg nodes to a dictionary of the parameter
+        name to type annotation as a string (for artifact annotations only).
+
+        Args:
+            args: AST argument nodes for a function.
+
+        Returns:
+            A dictionary of parameter name to type annotation as a string.
+        """
+        arg_to_ann = {}
+        for arg in args:
+            annotation = arg.annotation
+            # Handle InputPath() and OutputPath()
+            if isinstance(annotation, ast.Call):
```

Comment:
Don't quite follow the code, would it falls into this condition if users mistakenly use `Input(...)`, and what if it's some other unexpected calls like `PipelineParam(...)`?

Reply:
Discussed offline and removed handling for `InputPath` and `OutputPath`.

I agree that we should program more defensively around unexpected inputs, though I think this belongs in the `kfp.components.extract_component_interface` function. The issue of liberal acceptance of annotations is an issue that precedes this PR. WDYT about this as a separate work item?

Reply:
> though I think this belongs in the `kfp.components.extract_component_interface` function

I think you're probably right. I recall we do check and make sure the type annotation is one of our known usage, otherwise there will be compilation errors. As long as this change doesn't introduce regression to this UX, which I now think it probably doesn't, we should be all good.